homeassistant/components/switch/xiaomi_ir.py

```diff
@@ -0,0 +1,196 @@
+"""
+Support for the Xiaomi IR Remote (Chuangmi IR).
+
+For more details about this platform, please refer to the documentation
+https://home-assistant.io/components/ir_remote.xiaomi_miio/
+"""
+import asyncio
+import logging
+
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.components.switch import (
+    DOMAIN, PLATFORM_SCHEMA, SwitchDevice)
+from homeassistant.const import (
+    CONF_COMMAND_OFF, CONF_COMMAND_ON, CONF_FRIENDLY_NAME,
+    CONF_SWITCHES, CONF_HOST, CONF_TOKEN, CONF_SLOT, CONF_TIMEOUT)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util.dt import utcnow
+
+REQUIREMENTS = ['python-miio==0.3.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+SERVICE_LEARN = 'xiaomi_miio_learn_command'
+SERVICE_SEND = 'xiaomi_miio_send_command'
+
+DEFAULT_TIMEOUT = 10
+DEFAULT_SLOT = 1
+
+SWITCH_SCHEMA = vol.Schema({
+    vol.Optional(CONF_COMMAND_OFF, default=None): cv.string,
+    vol.Optional(CONF_COMMAND_ON, default=None): cv.string,
+    vol.Optional(CONF_FRIENDLY_NAME): cv.string,
+})
+
+LEARN_COMMAND_SCHEMA = vol.Schema({
+    vol.Optional("timeout", default=10):
+        vol.All(int, vol.Range(min=0)),
+    vol.Optional("slot", default=1):
+        vol.All(int, vol.Range(min=1, max=1000000)),
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Required(CONF_TOKEN): vol.All(str, vol.Length(min=32, max=32)),
+    vol.Optional(CONF_SWITCHES, default={}):
+        vol.Schema({cv.slug: SWITCH_SCHEMA}),
+}, extra=vol.ALLOW_EXTRA).extend(LEARN_COMMAND_SCHEMA.schema)
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Xiaomi IR Remote (Chuangmi IR) platform."""
+    from miio import ChuangmiIr, DeviceException
+    from construct import ChecksumError
```

Comment:
When can `ChecksumError` happen and should not this be something to handle inside python-miio?

Reply:
It happens if the token is the correct length but wrong.

Reply:
I think python-miio should be patched to emit `DeviceException` or (`IncorrectTokenException` which inherits from `DeviceException`), which should then be catched here.